### ISSUE TITLE
❓ [QUESTION] How exactly does the convolution work?

### ISSUE THREAD

**Issue by tueboesen** [e3nn/e3nn#264]
I have been trying to understand the convolution class for a while now, but I'm still not really sure how it works, so I was hoping that you could provide a few more details (I know that the convolution is an 'official' convolution but rather just one you guys have written in an example). The initialization of the class is rather well explained, but the forward function has no explanation, so I'm struggling a bit here.

```
@compile_mode('script')
class Convolution(torch.nn.Module):
    r"""equivariant convolution
    Parameters
    ----------
    irreps_in : `Irreps`
        representation of the input node features
    irreps_node_attr : `Irreps`
        representation of the node attributes
    irreps_edge_attr : `Irreps`
        representation of the edge attributes
    irreps_out : `Irreps` or None
        representation of the output node features
    number_of_edge_features : int
        number of scalar (0e) features of the edge used to feed the FC network
    radial_layers : int
        number of hidden layers in the radial fully connected network
    radial_neurons : int
        number of neurons in the hidden layers of the radial fully connected network
    num_neighbors : float
        typical number of nodes convolved over
    """
    def __init__(
        self,
        irreps_in,
        irreps_node_attr,
        irreps_edge_attr,
        irreps_out,
        number_of_edge_features,
        radial_layers,
        radial_neurons,
        num_neighbors
    ) -> None:
        super().__init__()
        self.irreps_in = o3.Irreps(irreps_in)
        self.irreps_node_attr = o3.Irreps(irreps_node_attr)
        self.irreps_edge_attr = o3.Irreps(irreps_edge_attr)
        self.irreps_out = o3.Irreps(irreps_out)
        self.num_neighbors = num_neighbors

        self.sc = FullyConnectedTensorProduct(self.irreps_in, self.irreps_node_attr, self.irreps_out)

        self.lin1 = FullyConnectedTensorProduct(self.irreps_in, self.irreps_node_attr, self.irreps_in)

        irreps_mid = []
        instructions = []
        for i, (mul, ir_in) in enumerate(self.irreps_in):
            for j, (_, ir_edge) in enumerate(self.irreps_edge_attr):
                for ir_out in ir_in * ir_edge:
                    if ir_out in self.irreps_out:
                        k = len(irreps_mid)
                        irreps_mid.append((mul, ir_out))
                        instructions.append((i, j, k, 'uvu', True))
        irreps_mid = o3.Irreps(irreps_mid)
        irreps_mid, p, _ = irreps_mid.sort()

        instructions = [
            (i_1, i_2, p[i_out], mode, train)
            for i_1, i_2, i_out, mode, train in instructions
        ]

        tp = TensorProduct(
            self.irreps_in,
            self.irreps_edge_attr,
            irreps_mid,
            instructions,
            internal_weights=False,
            shared_weights=False,
        )
        self.fc = FullyConnectedNet(
            [number_of_edge_features] + radial_layers * [radial_neurons] + [tp.weight_numel],
            torch.nn.functional.silu
        )
        self.tp = tp

        self.lin2 = FullyConnectedTensorProduct(irreps_mid, self.irreps_node_attr, self.irreps_out)

    def forward(self, node_input, node_attr, edge_src, edge_dst, edge_attr, edge_features) -> torch.Tensor:
        weight = self.fc(edge_features)

        x = node_input

        s = self.sc(x, node_attr)
        x = self.lin1(x, node_attr)

        edge_features = self.tp(x[edge_src], edge_attr, weight)
        x = scatter(edge_features, edge_dst, dim=0, dim_size=x.shape[0]).div(self.num_neighbors**0.5)

        x = self.lin2(x, node_attr)

        c_s, c_x = math.sin(math.pi / 8), math.cos(math.pi / 8)
        m = self.sc.output_mask
        c_x = (1 - m) + c_x * m
        return c_s * s + c_x * x

```
Let me start with the input to the forward routine, node_input I assume is actually node_features, just like you have edge_features, and these makes good sense. node_attr/edge_attr I'm a bit more uncertain on what normally goes in there, and how it is used.

Next confusion is the overall purpose of this convolution, I get what a convolution is normally, but I'm not sure I understand the strategy here. You have a bunch of tensorproducts, that mixes the various components of the node_features, but only among individual nodes (Unless I'm mistaken tensorproducts does not mix different nodes' features)
So the actual mixing of nodes/edges happens in this part I'm guessing? 
```
        edge_features = self.tp(x[edge_src], edge_attr, weight)
        x = scatter(edge_features, edge_dst, dim=0, dim_size=x.shape[0]).div(self.num_neighbors**0.5)
```
This is still a tensorproduct, but what I think happens is that the you convert the nodes features into edges edge features, and then in the scatter routine you essentially sum all the edge features back onto the nodes, which essentially allows you to mix information from neighbouring nodes. 

Now for the real confusing part:
```
        c_s, c_x = math.sin(math.pi / 8), math.cos(math.pi / 8)
        m = self.sc.output_mask
        c_x = (1 - m) + c_x * m
        return c_s * s + c_x * x
```
Here I'm really not sure what is going on, s is a mixing of the original node features, while x is the now convolved node features, but what is this self.sc.output_mask and how does that work exactly? and why are you mixing the unconvolved and convolved in this way?

**Comment by mariogeiger** [comment]
> Let me start with the input to the forward routine, node_input I assume is actually node_features, just like you have edge_features, and these makes good sense. node_attr/edge_attr I'm a bit more uncertain on what normally goes in there, and how it is used.

Sorry for the poor variable names.
```
node_input = node's feature data that flow in the network with message passing
node_attr = node's features that are independent of the message passing step (like atom type)
edge_attr = edge's features that are independent of the message passing step (like sh of the relative position vector)
edge_features (that I now renamed edge_scalars) = scalar quantities that are independent of the message passing step used to feed the MLP that generates the weights for the TP (usually function of the relative distance)
```

> So the actual mixing of nodes/edges happens in this part I'm guessing?

Exact

The reasons why we have other stuff are
1. self connections (`s` and the mix at the end)
2. [Depth-wise](https://arxiv.org/abs/1910.09455) inspired decomposition (`self.lin1` and `self.lin2`)

> what is this self.sc.output_mask and how does that work exactly? 

It's a mask (it contains 0s and 1s) that tells you which outputs (of here `self.sc` operation) are non-zero.
For instance `Linear("10x0e", "2x0e + 1e").output_mask` is `[1,1,0,0,0]`.

> and why are you mixing the unconvolved and convolved in this way?

The motivation is to have skip connections... when you can (`output_mask` tells you when you can)

When `output_mask = 0` you get `return c_s * 0 + 1 * x` (simply no skip connection)
When `output_mask = 1` you get `return sin(pi/8) * s + cos(pi/8) * x`

Why `sin(phi), cos(phi)`? Because `sin^2 + cos^2 = 1` which ensure a proper addition of variance.
Why `pi/8`? Because it's between 0 and pi/4 => it puts more weight on the skip connection, which is known to help training.

**Comment by tueboesen** [comment]
Thank you for the detailed explanation! 

I didn't realize that the variable named sc was named because it was referring to a skip connection, but now that you mention it is of course seems kinda obvious. 

> Why pi/8? Because it's between 0 and pi/4 => it puts more weight on the skip connection, which is known to help training.
Doesn't pi/8 put less weight on the skip connection than the convoluted connection or am I mixing things up here?

**Comment by mariogeiger** [comment]
Oh you are right. sin and cos are not in the order I had in mind. I can't tell if it was a typo when I wrote the code or if I'm influenced by the [last version](https://github.com/e3nn/e3nn/blob/main/e3nn/nn/models/v2103/points_convolution.py#L98-L104) of the code I wrote.

**Comment by tueboesen** [comment]
Cheers!
